### PR TITLE
fix: avoid config.yaml modification in standalone mode due to read only file system.

### PR DESCRIPTION
**Pull request by TheShubhendra**
- Removed modification of config.yaml in standalone mode.
- Added checks to make sure it uses data plane with yaml config provider.
- Added separate config file for standalone mode example.

### CODE REVIEW

**Comment by TheShubhendra** [comment]
Can you please review @kayx23

**Comment by shreemaan-abhishek** [comment]
the CI is not running, please push an empty commit

**Comment by TheShubhendra** [comment]
> the CI is not running, please push an empty commit

@shreemaan-abhishek I have pushed an empty commit but CI is still not running.

**Comment by TheShubhendra** [comment]
I had tested workflow `apisix dashboard all in one docker` on my local system, it works fine after  [#ef86c63](https://github.com/apache/apisix-docker/pull/558/commits/ef86c632c8d1c4f8cefabb6793e4072b858b9f1f) .
So I had modified the workflow file in my fork that shows it is not able to run the container to due segmentation fault as mentioned in this job run https://github.com/TheShubhendra/apisix-docker/actions/runs/10062135660/job/27813891259 .

**Comment by TheShubhendra** [comment]
Can someone please review?

**Comment by shreemaan-abhishek** [comment]
@TheShubhendra, please check why is the dashboard CI failing. Thank you for your patience.

**Comment by TheShubhendra** [comment]
> @TheShubhendra, please check why is the dashboard CI failing. Thank you for your patience.

As I mentioned before, the same workflow works on my PC, but it gives a segmentation fault on GitHub Actions. https://github.com/TheShubhendra/apisix-docker/actions/runs/10062135660/job/27813891259#:~:text=Segmentation%20fault%20(core%20dumped) This could be due to GitHub's resource limits or possibly another issue not related to apisix.

**Comment by kayx23** [comment]
Re-running failed jobs.
image

**Comment by TheShubhendra** [comment]
image


This issue seems to be from the main APISIX source code, not this repository.

**Comment by shreemaan-abhishek** [comment]
@TheShubhendra, thanks for the [fix](https://github.com/apache/apisix-docker/pull/558/commits/dca6df51ccda05840183b2ac60237c865463f8fc). Please make this fix in another PR and then pull changes from master when the fix PR is merged. One PR for one thing. Thanks for your patience.

**Comment by shreemaan-abhishek** [comment]
@TheShubhendra, please merge master branch to pull the fix.

File: redhat/docker-entrypoint.sh [hunk 1]
```diff
@@ -31,10 +31,29 @@ deployment:
     config_provider: yaml
 _EOC_
       else
-          # updating config.yaml for standalone mode
-          echo "$(sed 's/role: traditional/role: data_plane/g' ${PREFIX}/conf/config.yaml)" > ${PREFIX}/conf/config.yaml
-          echo "$(sed 's/role_traditional:/role_data_plane:/g' ${PREFIX}/conf/config.yaml)" > ${PREFIX}/conf/config.yaml
-          echo "$(sed 's/config_provider: etcd/config_provider: yaml/g' ${PREFIX}/conf/config.yaml)" > ${PREFIX}/conf/config.yaml
+          # Check if the deployment role is set to data_plane and config provider is set to yaml for standalone mode
```
Comment (shreemaan-abhishek) [inline redhat/docker-entrypoint.sh:34]: this part of code is being repeated across entrypoint scripts. Please DRY.

Comment (TheShubhendra) [reply redhat/docker-entrypoint.sh:34]: Thank you for your suggestion. What we can do is extract the logic into a single file and place it in the root directory. Then, we can modify the Makefile to copy this file into the appropriate build context directories (e.g., /debian, /redhat, and /debian-dev) before the Docker build process, and remove it afterward.

Please share your thoughts on this.

Comment (TheShubhendra) [reply redhat/docker-entrypoint.sh:34]:
```bash
#!/bin/bash

check_config() {
    local key="$1"
    local expected_value="$2"
    local error_message="$3"

    if ! grep -q "${key}: ${expected_value}" "${PREFIX}/conf/config.yaml"; then
        echo "Error: ${PREFIX}/conf/config.yaml does not contain '${key}: ${expected_value}'. ${error_message}"
        echo "Please refer to the APISIX documentation for deployment modes: https://apisix.apache.org/docs/apisix/deployment-modes/"
        exit 1
    fi
}

check_config "role" "data_plane" "Deployment role must be set to 'data_plane' for standalone mode."
check_config "role_data_plane" "" "Role must be set to 'role_data_plane:' for standalone mode."
check_config "config_provider" "yaml" "Config provider must be set to 'yaml' for standalone mode."
```
This file can be placed into /utils and then sourced into all entrypoint files.

Comment (shreemaan-abhishek) [reply redhat/docker-entrypoint.sh:34]: yes

**Review by shreemaan-abhishek — APPROVED** [review]

**Comment by TheShubhendra** [comment]
can someone please review this PR

**Comment by TheShubhendra** [comment]
Any updates on this?

**Comment by hiddenmarten** [comment]
I would like to vote on this PR, it's really helpful and solves a lot of confusing issues in the current repository state.

**Pull request merged**